cpuminer.go

```diff
@@ -496,6 +497,107 @@ func (m *CPUMiner) NumWorkers() int32 {
 	return int32(m.numWorkers)
 }
 
+// generateNBlocks generates the requested number of blocks. It is self
+// contained in that it creates block templates and attempts to solve them while
+// detecting when it is performing stale work and reacting accordingly by
+// generating a new block template.  When a block is solved, it is submitted.
+// The function returns a list of the hashes of generated blocks.
+func (m *CPUMiner) generateNBlocks(n uint32) ([]*wire.ShaHash, error) {
+	m.Lock()
+
+	// Respond with an error if there's virtually 0 chance of CPU-mining a block.
+	if !m.server.chainParams.GenerateSupported {
+		m.Unlock()
+		return nil, errors.New("No support for `generate` on the current " +
+			"network, " + m.server.chainParams.Net.String() +
+			", as it's unlikely to be possible to CPU-mine a block.")
+	}
+
+	// Respond with an error if server is already mining.
+	if m.started {
+		m.Unlock()
+		return nil, errors.New("Server is already CPU mining. Please call " +
+			"`setgenerate 0` before calling discrete `generate` commands.")
+	}
+
+	m.started = true
+
+	m.speedMonitorQuit = make(chan struct{})
+	m.wg.Add(1)
+	go m.speedMonitor()
+
+	m.Unlock()
+
+	minrLog.Tracef("Generating %d blocks", n)
+
+	i := uint32(0)
+	blockHashes := make([]*wire.ShaHash, n, n)
+
+	// Start a ticker which is used to signal checks for stale work and
+	// updates to the speed monitor.
+	ticker := time.NewTicker(time.Second * hashUpdateSecs)
+	defer ticker.Stop()
+
+	for {
+		// Read updateNumWorkers in case someone tries a `setgenerate` while
+		// we're generating. We can ignore it as the `generate` RPC call only
+		// uses 1 worker.
+		select {
+		case <-m.updateNumWorkers:
+		default:
+		}
+
+		// No point in searching for a solution before the chain is
+		// synced.  Also, grab the same lock as used for block
+		// submission, since the current block will be changing and
+		// this would otherwise end up building a new block template on
+		// a block that is in the process of becoming stale.
+		m.submitBlockLock.Lock()
+		_, curHeight := m.server.blockManager.chainState.Best()
+		if curHeight != 0 && !m.server.blockManager.IsCurrent() {
```

Comment:
Like the `ConnectedCount` check you already removed, I believe this check can go away too.  It makes sense for the normal mining path, but I don't think it should be checked for the discreet generate path.  What will happen is if you have your simnet node with new blocks for ~24 hours, you'd get stuck here since there are no other blocks coming in and this condition can never bet met.